Bluetooth: Host: Add more <err>-to_str() APIs

rugeGerritsen opened the pull request:
This can be useful if application developers want to print them in the applications.
Later we can also use them in the host to improve debuggability.

rugeGerritsen committed:
Bluetooth: Host: Define bt_att_err_to_str()

This can be useful if application developers
want to print them in the applications.

Later we can also use them in
the host to improve debuggability.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen committed:
Bluetooth: Host: Define bt_smp_err_to_str()

This API converts a SMP error code to a string.
This can be useful if application developers want
to print them in the applications.

BT_SMP_ERR_SUCCESS was added for completeness.

Later we can also use them in the host to improve debuggability.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -185,6 +185,59 @@ static struct bt_att_tx_meta_data tx_meta_data_storage[CONFIG_BT_ATT_TX_COUNT];
+
+const char *bt_att_err_to_str(uint8_t att_err)
+{
+	#define ATT_ERR(err) [err] = #err
+
+	/* To mapping tables are used to avoid a big gap with NULL-entries. */
+
+	const char * const first_mapping_table[] = {
+		ATT_ERR(BT_ATT_ERR_SUCCESS),
+		ATT_ERR(BT_ATT_ERR_INVALID_HANDLE),
+		ATT_ERR(BT_ATT_ERR_READ_NOT_PERMITTED),
+		ATT_ERR(BT_ATT_ERR_WRITE_NOT_PERMITTED),
+		ATT_ERR(BT_ATT_ERR_INVALID_PDU),
+		ATT_ERR(BT_ATT_ERR_AUTHENTICATION),
+		ATT_ERR(BT_ATT_ERR_NOT_SUPPORTED),
+		ATT_ERR(BT_ATT_ERR_INVALID_OFFSET),
+		ATT_ERR(BT_ATT_ERR_AUTHORIZATION),
+		ATT_ERR(BT_ATT_ERR_PREPARE_QUEUE_FULL),
+		ATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_FOUND),
+		ATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_LONG),
+		ATT_ERR(BT_ATT_ERR_ENCRYPTION_KEY_SIZE),
+		ATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN),
+		ATT_ERR(BT_ATT_ERR_UNLIKELY),
+		ATT_ERR(BT_ATT_ERR_INSUFFICIENT_ENCRYPTION),
+		ATT_ERR(BT_ATT_ERR_UNSUPPORTED_GROUP_TYPE),
+		ATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES),
+		ATT_ERR(BT_ATT_ERR_DB_OUT_OF_SYNC),
+		ATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED),
+	};
+
+	#define ATT_ERR_SECOND(err) [err - BT_ATT_ERR_WRITE_REQ_REJECTED] = #err
```
```suggestion
	#define ATT_ERR(err) [err] = #err
	#define ATT_ERR_SECOND(err) [err - BT_ATT_ERR_WRITE_REQ_REJECTED] = #err

	/* To mapping tables are used to avoid a big gap with NULL-entries. */

	const char * const first_mapping_table[] = {
		ATT_ERR(BT_ATT_ERR_SUCCESS),
		ATT_ERR(BT_ATT_ERR_INVALID_HANDLE),
		ATT_ERR(BT_ATT_ERR_READ_NOT_PERMITTED),
		ATT_ERR(BT_ATT_ERR_WRITE_NOT_PERMITTED),
		ATT_ERR(BT_ATT_ERR_INVALID_PDU),
		ATT_ERR(BT_ATT_ERR_AUTHENTICATION),
		ATT_ERR(BT_ATT_ERR_NOT_SUPPORTED),
		ATT_ERR(BT_ATT_ERR_INVALID_OFFSET),
		ATT_ERR(BT_ATT_ERR_AUTHORIZATION),
		ATT_ERR(BT_ATT_ERR_PREPARE_QUEUE_FULL),
		ATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_FOUND),
		ATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_LONG),
		ATT_ERR(BT_ATT_ERR_ENCRYPTION_KEY_SIZE),
		ATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN),
		ATT_ERR(BT_ATT_ERR_UNLIKELY),
		ATT_ERR(BT_ATT_ERR_INSUFFICIENT_ENCRYPTION),
		ATT_ERR(BT_ATT_ERR_UNSUPPORTED_GROUP_TYPE),
		ATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES),
		ATT_ERR(BT_ATT_ERR_DB_OUT_OF_SYNC),
		ATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED),
	};
```

Since they are being undef'ed together, might as well define them together :)

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -185,6 +185,59 @@ static struct bt_att_tx_meta_data tx_meta_data_storage[CONFIG_BT_ATT_TX_COUNT];
+	if (att_err < ARRAY_SIZE(first_mapping_table) && first_mapping_table[att_err]) {
+		return first_mapping_table[att_err];
+	} else if (att_err >= BT_ATT_ERR_WRITE_REQ_REJECTED &&
+		   att_err - BT_ATT_ERR_WRITE_REQ_REJECTED < ARRAY_SIZE(second_mapping_table) &&
+		   second_mapping_table[att_err - BT_ATT_ERR_WRITE_REQ_REJECTED]) {
+		return second_mapping_table[att_err - BT_ATT_ERR_WRITE_REQ_REJECTED];
```
Suggest to store `att_err - BT_ATT_ERR_WRITE_REQ_REJECTED` in a local variable to make this easier to read

Thalley commented on tests/bluetooth/gatt/src/main.c:
What about unittests for the SMP ones?

rugeGerritsen replied:
Do you have a suggestion where to put them?

Thalley replied:
Probably requires a new unit test directory :( 

You could also argue that since this is testing ATT error codes, it should not be in the `gatt` directory

rugeGerritsen replied:
Yeah, I noticed. Do you have an example unit test which I can take as a starting point? Then I can create new unit tests for both the new APIs.

Not sure where to put the SMP thing though as the API is simply defined in conn.h

Thalley replied:
> Yeah, I noticed. Do you have an example unit test which I can take as a starting point? Then I can create new unit tests for both the new APIs.

The GATT unit test directory seems to be pretty decent for this. 

> Not sure where to put the SMP thing though as the API is simply defined in conn.h

Since unittests are not necessarily API tests, I think it makes sense to just create a SMP directory.

Thalley commented on include/zephyr/bluetooth/att.h:
```diff
@@ -101,6 +101,20 @@ extern "C" {
 
+/** Converts a ATT error to string.
+ *
+ * The error codes are described in the Bluetooth Core specification,
+ * Vol 3, Part F, Section 3.4.1.1.
```
The "second" ones are from the core supplemental spec

Thalley reviewed (changes requested):
Great! I think we will attempt this "template" for the to_str functions in LE Audio as well

Thalley commented on include/zephyr/bluetooth/att.h:
```diff
@@ -101,6 +101,20 @@ extern "C" {
+ *
+ * @return The string representation of the ATT error code.
+ */
+const char *bt_att_err_to_str(uint8_t att_err);
```
Does it make sense to also have `bt_gatt_err_to_str` which basically just calls this with `-att_err`?

rugeGerritsen replied:
Yeah, that sounds like a good idea. That may help some users not knowing the difference between ATT/GATT

Thalley replied:
Indeed, and most users at working at the GATT level. 

The GATT function may want to assume that users are unaware of what they are doing do be something like 
```c
const char *bt_gatt_err_to_str(uint8_t gatt_err)
{
    bt_att_err_to_str(gatt_err < 0 ? -gatt_err : gatt_err)
}
```

so that it's works for both e.g. `-2` and `2` for `BT_ATT_ERR_READ_NOT_PERMITTED`

Thalley replied:
The biggest challenge there is of course having users know the difference between GATT errors and errno values...

rugeGerritsen committed:
Bluetooth: Host: Define bt_gatt_err_to_str()

The function reuses the ATT implementation.
To make the function simpler to use, the function handles both positive
and negative values.

Unfortunately the APIs do not document if the API returns an
errno val or a GATT return value.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -185,6 +185,59 @@ static struct bt_att_tx_meta_data tx_meta_data_storage[CONFIG_BT_ATT_TX_COUNT];
+		ATT_ERR_SECOND(BT_ATT_ERR_OUT_OF_RANGE),
+	};
+
+	const uint8_t second_index = att_err - BT_ATT_ERR_WRITE_REQ_REJECTED;
+
+	if (att_err < ARRAY_SIZE(first_mapping_table) && first_mapping_table[att_err]) {
+		return first_mapping_table[att_err];
+	} else if (att_err >= BT_ATT_ERR_WRITE_REQ_REJECTED &&
+		   second_index < ARRAY_SIZE(second_mapping_table) &&
+		   second_mapping_table[second_index]) {
+		return second_mapping_table[second_index];
```
```suggestion
	if (att_err < ARRAY_SIZE(first_mapping_table) && first_mapping_table[att_err]) {
		return first_mapping_table[att_err];
	} else if (att_err >= BT_ATT_ERR_WRITE_REQ_REJECTED) {
		const uint8_t second_index = att_err - BT_ATT_ERR_WRITE_REQ_REJECTED;
		
		if (second_index < ARRAY_SIZE(second_mapping_table) &&
		    second_mapping_table[second_index]) {
			return second_mapping_table[second_index];
		}
```

To avoid unnecessary calculations and to avoid underflowing `att_err`

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1828,6 +1828,20 @@ int bt_conn_auth_pairing_confirm(struct bt_conn *conn);
  */
 int bt_conn_auth_pincode_entry(struct bt_conn *conn, const char *pin);
 
+/** Converts a SMP error to string.
+ *
+ * The error codes are described in the Bluetooth Core specification,
+ * Vol 3, Part H, Section 3.5.5.
+ *
+ * The Security Manager Protocol documentation found in Vol 4, Part H,
+ * describes when the different error codes are used.
+ *
+ * See also the defined BT_SMP_ERR_* macros.
+ *
+ * @return The string representation of the SMP error code.
+ */
+const char *bt_smp_err_to_str(uint8_t smp_err);
```
Should the error codes be moved to a public place? They are currently only defined in an internal header file, which actually makes this function useless :) 

Or did you mean to use this for the `enum bt_security_err` which is the public representation of the SMP error codes?

rugeGerritsen replied:
I thought the same - the SMP error codes should be added to a public header. We should also add an API to convert `enum bt_security_err` to string

Thalley replied:
But is the SMP error codes ever provided to the application? 
If not, then there's no point in providing a public function for a set of error values that never reach the public API. 

Perhaps the SMP converting function should be internal and just expose a public function for the `bt_security_err`.

rugeGerritsen replied:
I agree. We should not add this API as a public API

Thalley commented on subsys/bluetooth/host/smp.h:
```diff
@@ -178,3 +179,17 @@ int bt_smp_sign(struct bt_conn *conn, struct net_buf *buf);
 
 /** Generate IRK from Identity Root (IR) */
 int bt_smp_irk_get(uint8_t *ir, uint8_t *irk);
+
+/** Converts a SMP error to string.
+ *
+ * The error codes are described in the Bluetooth Core specification,
+ * Vol 3, Part H, Section 3.5.5.
+ *
+ * The Security Manager Protocol documentation found in Vol 4, Part H,
+ * describes when the different error codes are used.
+ *
+ * See also the defined BT_SMP_ERR_* macros.
+ *
+ * @return The string representation of the SMP error code.
+ */
+const char *bt_smp_err_to_str(uint8_t smp_err);
```
Since this is now an internal function, are you planning on using it in the host somewhere?

rugeGerritsen replied:
This seems like a good place to use it: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/smp.c#L1653

Thalley replied:
Indeed :)

rugeGerritsen committed:
Bluetooth: Host: Define bt_security_err_to_str()

This can be useful if application developers
want to print them in the applications.

Later we can also use them in
the host to improve debuggability.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen commented:
I've added the 3.7.0 milestone because: 
* It is a low-risk PR that doesn't change existing functionality, it just adds new, standalone, tested `to_str()` APIs.
